lib/common/reverse_tunnel.c

```diff
@@ -0,0 +1,189 @@
+/*
+ * Copyright (c) 2024 Ichito Nagata, Fastly, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to
+ * deal in the Software without restriction, including without limitation the
+ * rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
+ * sell copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+ * IN THE SOFTWARE.
+ */
+
+#include "h2o/reverse_tunnel.h"
+#include "h2o.h"
+
+static inline void schedule_reconnect(h2o_reverse_ctx_t *reverse)
+{
+    h2o_timer_link(reverse->accept_ctx->ctx->loop, reverse->config.reconnect_interval, &reverse->reconnect_timer);
+}
+
+struct on_reverse_close_data_t {
+    void (*orig_cb)(void *data);
+    void *orig_data;
+    h2o_reverse_ctx_t *reverse;
+};
+
+void on_reverse_close(void *_data) {
+    struct on_reverse_close_data_t *data = (void *)_data;
+    data->orig_cb(data->orig_data);
+
+    h2o_reverse_ctx_t *reverse = data->reverse;
+    reverse->httpclient.client = NULL;
+    h2o_mem_clear_pool(&reverse->pool);
+
+    schedule_reconnect(reverse);
+}
+
+static void on_reverse_reconnect_timeout(h2o_timer_t *timer)
+{
+    h2o_reverse_ctx_t *reverse = H2O_STRUCT_FROM_MEMBER(h2o_reverse_ctx_t, reconnect_timer, timer);
+    h2o_reverse_start_listening(reverse);
+}
+
+static void on_reverse_read(h2o_socket_t *sock, const char *err)
+{
+    h2o_reverse_ctx_t *reverse = (void *)sock->data;
+
+    h2o_socket_read_stop(sock);
+    if (err != NULL) {
+        h2o_error_printf("unexpected read error in on_reverse_rea: %s\n", err);
+        return;
+    }
+
+    h2o_accept(reverse->accept_ctx, sock);
```

Comment:
> we probably want to call h2o_http1_accept or h2o_http2_accept directly from here,

Thanks, agree with this

> depending on the value of the ALPN request header field

Is it `Selected-ALPN` response header field?

> also, we have to make the decision when we send the response headers, including the ALPN response header to indicate the protocol being chosen

I totally forgot about ALPN **request** header, and am willing to add it
NOTE: again, this is reverse-tunnel **client** implementation, not server.

Reply:
> I'm not sure if that is the correct thing to do with protocols like HTTP/2 that allows both endpoints start talking at the same time.
I think, in the case of http2, we may have to do something similar to `on_upgrade_complete` in lib/http2/connection.c, right?

> Maybe it is a sign that the approach is incorrect.

Instead of doing this in the socket read callback, can we do it in on_head and return a signal with semantics like `h2o_httpclient_socket_has_been_already_stolen_do_not_touch_it_anymore`?

Reply:
> I think, in the case of http2, we may have to do something similar to `on_upgrade_complete` in lib/http2/connection.c, right?

I assume you are pointing to the fact that in `on_upgrade_complete` we are consuming the bytes in the receive side and building HTTP/1.1 response headers of our own. Yes, that is one way forward.

> Instead of doing this in the socket read callback, can we do it in on_head and return a signal with semantics like `h2o_httpclient_socket_has_been_already_stolen_do_not_touch_it_anymore`?

First of all, I still think that it is correct to return a special value. I assume you think the same. Then, the question is if:
* we want to pass the size of the response header to the `on_head` callback so that it can steal the socket immediately, or
* we want to let the `on_head` set a callback (inside `h2o_httpclient_properties_t`) that will be invoked by httpclient once httpclient cosumes the response headers.

I lean towards the former.

Reply:
Another option would be: passing `h2o_socket_t *(*steal_socket)(void)` function pointer as a field of h2o_httpclient_properties, and the consumer calls it if it want to steal the socket. By doing so we can write buffer consuming code in it, and as an additional bonus, we no longer need `h2o_httpclient_steal_socket` return value to signal. Of course it can be a public function like `h2o_socket_t *h2o_httpclient_steal_socket(void)` but it might be a bit complex to dispatch to the actual client implementations (h1/h2/h3).

Reply:
Yeah, I think I prefer returning a special value, as that would make things more explicit in the source code of http1client.c. We _need_ to return something, therefore the cost of having a signal is marginal.

What I am concerned is that pointers to the response being retained by http1client becomes invalid the moment the socket receive buffer is consumed. If that is going to happen at a different moment for stolen socket, that should be easy to understand when reading http1client.c. Otherwise, we might break things.

Reply:
ok then let's have `rlen` in `h2o_httpclient_conn_properties_t`.

By the way, just to confirm, another obvious solution is setting zero-timeout (deferred) callback in `on_head` and do all the things in that callback, but you don't like that, do you?

Reply:
Having rlen (or `size_t *rlen` so that the application cannot steal H2 and H3 connectsions) sounds good to me.

> By the way, just to confirm, another obvious solution is setting zero-timeout (deferred) callback in `on_head` and do all the things in that callback, but you don't like that, do you?

Yeah I think doing that just makes maintenance harder. I think what we want is a clear cut and explicit transition of ownership of the socket (from http1client to the app). The idea of using a deferred callback to give http1client the time to consumes the bytes is exactly the opposite.

Reply:
Just added `steal_bytes` to conn props and set rlen value in https://github.com/h2o/h2o/pull/3405/commits/216b991f6bbabed67653ef28dccc03e3f953e5b6